File: erpnextswiss/erpnextswiss/page/match_payments/match_payments.js
frappe.pages['match_payments'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __('Match payments'),
		single_column: true
	});

	frappe.match_payments.make(page);
	frappe.match_payments.run(page);
    
    // add the application reference
    frappe.breadcrumbs.add("ERPNextSwiss");
}

frappe.match_payments = {
	start: 0,
	make: function(page) {
		var me = frappe.match_payments;
		me.page = page;
		me.body = $('<div></div>').appendTo(me.page.main);
		var data = "";
		$(frappe.render_template('match_payments', data)).appendTo(me.body);

        // add menu button
        this.page.add_menu_item(__("Open bank import"), function() {
            // navigate to bank import tool
            window.location.href="/app/bankimport";
		});
        
		// attach button handlers
		this.page.main.find(".btn-match").on('click', function() {                   
            match(page);
		});
        
        this.page.main.find(".btn-auto-match-id").on('click', function() {                   
            auto_match(page, "docid");
		});

        this.page.main.find(".btn-auto-match-esr").on('click', function() {                   
            auto_match(page, "esr");
		});
	},
	run: function(page) {
		// read unpaid sales invoices
		frappe.call({
			method: 'erpnextswiss.erpnextswiss.page.match_payments.match_payments.get_open_sales_invoices',
			args: { },
			callback: function(r) {
				if (r.message) {
					var parent = page.main.find(".sales-invoice-table").empty();
                    if (r.message.unpaid_sales_invoices.length > 0) {
                        $(frappe.render_template('sales_invoice_table', r.message)).appendTo(parent);
                    } else {
                        $('<p class="text-muted">' + __("No unpaid sales invoices found.") + '</p>').appendTo(parent);
                    }
				} 
			}
		});
        // read unallocated payment entries
		frappe.call({
			method: 'erpnextswiss.erpnextswiss.page.match_payments.match_payments.get_unallocated_payment_entries',
			args: { },
			callback: function(r) {
				if (r.message) {
					var parent = page.main.find(".payment-table").empty();
                    if (r.message.unallocated_payment_entries.length > 0) {
                        $(frappe.render_template('payment_entry_table', r.message)).appendTo(parent);
                    } else {
                        $('<p class="text-muted">' + __("No unallocated payment entries found.") + '</p>').appendTo(parent);
                    }
				} 
			}
		});
	},
    start_wait: function() {
        document.getElementById("waitingScreen").style.display = "block";
    },
    end_wait: function() {
        document.getElementById("waitingScreen").style.display = "none";
    }
}

function match(page) {
    // get sales invoice
    try {
        var sales_invoice = document.querySelector('input[name="invoice"]:checked').value;
        
        try {
            // get payment entry
            var payment_entry = document.querySelector('input[name="payment"]:checked').value;

            //frappe.msgprint("Match " + sales_invoice + " with " + payment_entry);
                    
            // enable waiting gif
            frappe.match_payments.start_wait();
            
            // call match method 
            frappe.call({
                method: 'erpnextswiss.erpnextswiss.page.match_payments.match_payments.match',
                args: {
                    'sales_invoice': sales_invoice,
                    'payment_entry': payment_entry
                },
                callback: function(r) {
                    if (r.message) {
                        //frappe.msgprint("Matched!");
                        
                        submit(r.message.payment_entry, page);
                    } 
                }
            }); 
        }
        catch (err) {
             frappe.msgprint( __("Please select a payment entry.") );
        }
    }
    catch (err) {
        frappe.msgprint( __("Please select a sales invoice.") );
    }
}

function auto_match(page, method="docid") {
    //try {                   
        // enable waiting gif
        frappe.match_payments.start_wait();
        
        // call match method 
        frappe.call({
            method: 'erpnextswiss.erpnextswiss.page.match_payments.match_payments.auto_match',
            args: {
                'method': method
            },
            callback: function(r) {
                if (r.message) {   
                    if (r.message.payments.length > 0) {    
                        frappe.show_alert( __("Matched {0} sales invoices.").replace("{0}", String(r.message.payments.length)));      
                        submit_all(r.message.payments, page);
                    } else {
                        completed(page, false);
                        frappe.show_alert( __("No matches found.") );
                    }
                } 
            }
        }); 
    //}
    //catch (err) {
    //     frappe.msgprint( __("Please select a payment entry.") );
    //}
}

function submit(payment_entry, page) {
    frappe.call({
        method: 'erpnextswiss.erpnextswiss.page.match_payments.match_payments.submit',
        args: { 
                'payment_entry': payment_entry
            },
        callback: function(r) {
            if (r.message) {
                completed(page);
            } 
        }
    });
}

function submit_all(payments, page) {
    frappe.call({
        method: 'erpnextswiss.erpnextswiss.page.match_payments.match_payments.submit_all',
        args: { 
                'payment_entries': payments
            },
        callback: function(r) {
            if (r.message) {
                completed(page);
            } 
        }
    });
}

function completed(page, reload=true) {
    // disable waiting gif
    frappe.match_payments.end_wait();
    
    // refresh
    if (reload) {
        location.reload(); 
    }
}

File: erpnextswiss/erpnextswiss/page/match_payments/match_payments.py
# Copyright (c) 2017-2018, libracore and contributors
# License: AGPL v3. See LICENCE

from __future__ import unicode_literals
import frappe
from frappe import throw, _
from erpnextswiss.erpnextswiss.page.bankimport.bankimport import create_reference

@frappe.whitelist()
def get_open_sales_invoices():
    # get unpaid sales invoices
    sql_query = ("""SELECT `name`, `customer`, `base_grand_total`, `outstanding_amount`, `due_date`, `esr_reference` 
                FROM `tabSales Invoice` 
                WHERE `docstatus` = 1 AND `outstanding_amount` > 0 
                ORDER BY `due_date` ASC""")
    unpaid_sales_invoices = frappe.db.sql(sql_query, as_dict=True)
    return {'unpaid_sales_invoices': unpaid_sales_invoices }

@frappe.whitelist()
def get_unallocated_payment_entries():
    # get unallocated payment entries
    sql_query = ("""SELECT `name`, `party`, `paid_amount`, `posting_date`, `remarks` 
                FROM `tabPayment Entry` 
                WHERE `docstatus` = 0  
                  AND `payment_type` = 'Receive' 
                ORDER BY `posting_date` ASC""")
    unallocated_payment_entries = frappe.db.sql(sql_query, as_dict=True)
    return {'unallocated_payment_entries': unallocated_payment_entries }

@frappe.whitelist()
def match(sales_invoice, payment_entry):
    # get the customer
    customer = frappe.get_value("Sales Invoice", sales_invoice, "customer")
    if customer:
        payment_entry_record = frappe.get_doc("Payment Entry", payment_entry)
        if payment_entry:
            # assign the actual customer
            payment_entry_record.party = customer            
            payment_entry_record.save()
            
            # now, add the reference to the sales invoice
            create_reference(payment_entry, sales_invoice)
            
            # and finally submit the payment entry (direct submit will void the reference :-( )
            #payment_entry_record.submit()
            
            return { 'payment_entry': payment_entry_record.name }
        else:
            return { 'error': "Payment entry not found" }
    else:
        return { 'error': "Customer not found" }
 
@frappe.whitelist()
def submit(payment_entry):
    payment_entry_record = frappe.get_doc("Payment Entry", payment_entry)
    payment_entry_record.submit()
    return { 'message': "Done" }

@frappe.whitelist()
def submit_all(payment_entries):
    # get the array from the string parameter
    payments = eval(payment_entries)
    # loop through all matched payments and submit them
    for payment_entry in payments:
        submit(payment_entry)
    return { 'message': "Done" }

@frappe.whitelist()
def auto_match(method="docid"):
    # make method lower case
    method = method.lower()
    # prepare array of matched payments
    matched_payments = []
    # read all new payments
    new_payments = get_unallocated_payment_entries()['unallocated_payment_entries']
    # loop through all unpaid sales invoices
    pending_sales_invoices = get_open_sales_invoices()['unpaid_sales_invoices']
    for unpaid_sales_invoice in pending_sales_invoices:
        if method == "docid":
            for payment in new_payments:
                if unpaid_sales_invoice['name'] in payment['remarks']:
                    matched_payment_entry = match(unpaid_sales_invoice['name'], payment['name'])['payment_entry']
                    matched_payments.append(matched_payment_entry)
        elif method == "esr":
            # only check Sales Invoice records with an ESR reference
            if unpaid_sales_invoice['esr_reference']:
                for payment in new_payments:
                    if unpaid_sales_invoice['esr_reference'].replace(' ', '') in payment['remarks']:
                        matched_payment_entry = match(unpaid_sales_invoice['name'], payment['name'])['payment_entry']
                        matched_payments.append(matched_payment_entry)
    return { 'message': "Done", 'payments': matched_payments }
